### ISSUE TITLE
Migrate to java.time

### ISSUE THREAD

**Issue by w4** [elastic/logstash#10525]
Now that the Logstash API is exposed to us via Java and Java 11 support is here, it'd be nice to use the native java.time classes rather than having to use Joda.

**Comment by yaauie** [comment]
While I agree that it would be nice to remove one more dependency, I think a little more context may help to explain the motivation for the change you are requesting.

The Logstash API (as exposed by package `co.elastic.logstash.api`) uses `java.time.Instant` to represent time; is there something else that is causing you to interact with JodaTime objects?

**Comment by w4** [comment]
I'm returning a date from my codec so it ends up gets passed through the `Valuefier` which doesn't have any `java.time` converters. I'm having to pass a `org.logstash.Timestamp(org.joda.time.DateTime)` back at the moment.

It's not currently an issue - we're constructing our date using `java.time`-compatible APIs anyway but I thought I'd get this tracked so it doesn't come as a surprise to us when it inevitably happens!

**Comment by yaauie** [comment]
Ah. So `Valuefier` support for a `java.time.Instant` would get you sorted?

**Comment by w4** [comment]
Yep that'll do it, although is that going to throw away my `DateTime` object just to create it again in `Timestamp` or is it stored as the `Instant`? 

Sorry to pick at the implementation, I'm sort of going on a mad one optimising these Java plugins as we've had loads of GC issues with our Ruby ones

**Comment by yaauie** [comment]
>  my `DateTime` object

What `DateTime` object are you talking about? `java.time.ZonedDateTime`? Adding the `Valuefier` converters to accept any number of inputs would be trivial, but we would need to know which ones you need.

>  just to create it again in `Timestamp`

Adding converters to the `Valuefier` would not change the event's internal representation.

From your GC-optimising viewpoint, I am guessing that adding inputs to the `Valuefier` (while keeping the event's internal representation the same) will _not_ address your need; you seem to be asking us to change the internal representation to one that would allow your use-case to eliminate CG churn. This would be a significantly larger undertaking, and may not be possible without breaking changes, since the `Rubyfier` currently coerces timestamp-like objects to its wrapper that uses JodaTime under the hood and exposes those details via `getTimestamp()` etc.

**Comment by w4** [comment]
Ah apologies that's me not knowing how the `java.time` API works, we get back a `TemporalAccessor` from `DateTimeFormatter.parse` that we can then use to construct an `Instant` so yep that'll work for us (although I am a bit miffed over how lenient Java is with throwaway allocations in all these new java.time/streams/etc APIs but thats for another conversation 😅)

We do believe a fair bit of GC churn is caused by the context switching and the `Rubyfier`-cations that come with it but that's understandably way out scope to start rearchitecting (thank you all for your work on Javafication & working towards that though).

My concerns were mainly around how the `Instant` would be handled by the `Valuefier`—as I understand it a `org.logstash.Timestamp` will be instantiated for the Rubifier to wrap so I'm thinking as to whether it would make sense to have `Timestamp` act as an interface for Joda's `DateTime` or a `TemporalAccessor` or what have you that I can store in my event to avoid that allocation on every context switch. `Timestamp` does leak the Joda implementation at the moment though so may not be trivial & will be breaking.

We've managed to make our whole log parsing pipeline temporary allocationless and of course we do still have a GC so a few little `Timestamp` objects here and there aren't *actually* a problem, plus the JodaTime implementation currently in place works just fine so this is all definitely just a nice-to-have.
